src/domains.rs

```diff
@@ -124,53 +131,144 @@ pub mod types {
+        pub fn new(name: &str) -> Self {
+            Self {
+                name: name.to_owned(),
+                region: None,
+            }
+        }
+
+        /// Specifies the region for sending emails from the domain.
+        pub fn with_region(mut self, region: impl Into<Region>) -> Self {
+            self.region = Some(region.into());
+            self
+        }
+    }
+
+    /// Region where [`SendEmail`]s will be sent from.
     ///
     /// Possible values are 'us-east-1' | 'eu-west-1' | 'sa-east-1' | 'ap-northeast-1'.
-    // TODO: Custom region.
-    #[derive(Debug, Copy, Clone, Serialize)]
+    ///
+    /// [`SendEmail`]: crate::types::SendEmail
+    #[derive(Debug, Clone)]
     pub enum Region {
-        #[serde(rename = "us-east-1")]
+        /// 'us-east-1'
         UsEast1,
-        #[serde(rename = "eu-west-1")]
+        /// 'eu-west-1'
         EuWest1,
-        #[serde(rename = "sa-east-1")]
+        /// 'sa-east-1'
         SaEast1,
-        #[serde(rename = "ap-northeast-1")]
+        /// 'ap-northeast-1'
         ApNorthEast1,
+        /// Custom or unrecognized region.
+        Custom(String),
+    }
+
+    impl Region {
+        /// Creates a new custom [`Region`].
+        pub fn new(name: &str) -> Self {
+            Self::Custom(name.to_owned())
+        }
+    }
+
+    impl Serialize for Region {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+        {
+            let name = match self {
+                Self::UsEast1 => "us-east-1",
+                Self::EuWest1 => "eu-west-1",
+                Self::SaEast1 => "sa-east-1",
+                Self::ApNorthEast1 => "ap-northeast-1",
+                Self::Custom(x) => x.as_str(),
+            };
+
+            serializer.serialize_str(name)
+        }
+    }
+
+    impl<'de> Deserialize<'de> for Region {
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            struct RegionVisitor;
+
+            impl<'de> Visitor<'de> for RegionVisitor {
+                type Value = Region;
+
+                fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                    f.write_str("a valid region name")
+                }
+
+                fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+                where
+                    E: Error,
+                {
+                    let value = match v {
+                        "us-east-1" => Region::UsEast1,
+                        "eu-west-1" => Region::EuWest1,
+                        "sa-east-1" => Region::SaEast1,
+                        "ap-northeast-1" => Region::ApNorthEast1,
+                        x => Region::Custom(x.to_owned()),
+                    };
+
+                    Ok(value)
+                }
+            }
+
+            deserializer.deserialize_str(RegionVisitor)
+        }
     }
```

Comment:
For this wouldn't it possible to use `serde(rename)`s instead of implementing serializing and deserializing? It's been a while since I've used serde but wouldn't this work:

```rs
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Region {
    /// 'us-east-1'
    #[serde(rename = "us-east-1")]
    UsEast1,
    /// 'eu-west-1'
    #[serde(rename = "eu-west-1")]
    EuWest1,
    /// 'sa-east-1'
    #[serde(rename = "sa-east-1")]
    SaEast1,
    /// 'ap-northeast-1'
    #[serde(rename = "ap-northeast-1")]
    ApNorthEast1,
}
```

Also judging from the docs, I don't think there's a reason to have the `Custom(String)` entry, is this something the Go library does? Because [the docs](https://resend.com/docs/api-reference/domains/create-domain) specifically mention:

> Possible values: `us-east-1' | 'eu-west-1' | 'sa-east-1' | 'ap-northeast-1'`

Reply:
That's what It used to be, but how do you add `Custom(String)`?

Reply:
Considering that there's only 4 regions I think we can remove the `Custom` entry, it is trivial to add more regions to the enum when the API adds them

Reply:
Alright, reverted and added #[non_exhaustive].